### ISSUE TITLE
Add return codes to mouse.wait_for_click and keyboard.wait_for_keypress

### ISSUE THREAD

**Issue by josephj11** [autokey/autokey#659]
### Has this issue already been reported?

- [X] I have searched through the existing issues.

### Is this a question rather than an issue?

- [X] This is not a question.

### What type of issue is this?

Enhancement

### Which Linux distribution did you use?

N/A

### Which AutoKey version did you use?

N/A

### How did you install AutoKey?

N/A

### Can you briefly describe the issue?

I want to know how a wait for click or keypress completed so I can use that information for flow control in scripts.

E.g. a loop continues indefinitely until the mouse is clicked. This has to distinguish between a timeout and a click.
The same thing with a loop terminated by a keypress.

### Can the issue be reproduced?

N/A

### What are the steps to reproduce the issue?

N/A

### What should have happened?

These API calls should return 0 for success and one or more defined non-zero values to cover any alternatives. So far, 1 for timeout/failure is all that comes to mind.

### What actually happened?

AFAIK, they do not return any status code - which is equivalent to returning 0 no matter what happened.

**Comment by Elliria** [comment]
I can confirm this and agree that it would be really useful to get some returns that can be used to react to the mouse clicks.

**Comment by Elliria** [comment]
I hadn't noticed that ```keyboard.wait_for_keypress``` didn't have a return, either, but I just tested it and can confirm that one, too.

**Comment by sl5net** [comment]
@josephj11 seems it works already 

https://askubuntu.com/a/639337/1017209
https://github.com/autokey/autokey/issues/126

**Comment by josephj11** [comment]
@sl5net That's an interesting Q&A. Thanks. The proposed solution shouldn't work. However, the trick of seeing if the timeout occurred is a viable workaround for the function's deficiency. (The Q&A is too old for posting a followup.) 

It would still be cleaner to have the two functions return status codes as requested.

**Comment by sebastiansam55** [comment]
The documentation can definitely be updated but from my reading the code they're (both `wait_for_keypress` and `wait_for_keyevent`) returning `self.event.wait(self.time_out)`

this translates to return `False` if you hit the timeout, `True` otherwise

Is that not the behavior you're seeing?

**Comment by Elliria** [comment]
Nope. It's a [known issue](https://github.com/autokey/autokey/issues/712) and always returns **False**:
```python
retCode = keyboard.wait_for_keypress('a',modifiers=['<ctrl>'],timeOut=5)

if retCode:
	myMessage = 'Wait for keypress exit code was: ' + str(retCode)
	dialog.info_dialog(title='You pressed <ctrl>+a', message=myMessage, width='200')
else:
	myMessage = 'Wait for keypress exit code was: ' + str(retCode)
	dialog.info_dialog(title='Timeout', message=myMessage, width='200')
# The exit code is always False.
```
I guess the mouse and keyboard both need access to the return codes.

**Comment by sebastiansam55** [comment]
yeah, don't know what to tell you it returns `True` if I press the key it's waiting for and `False` if it hits the timeout, that's the expected behavior to me

using your script; 

you can see it return false with no input and then return true when I hit ctrl+a while in the timout period
![Screenshot from 2023-03-29 08-22-28](https://user-images.githubusercontent.com/1707320/228533975-7e3b77cc-fb7a-4e8e-a424-35ae50221c59.png)

**Comment by Elliria** [comment]
This is one of the rare times when I make a mistake and am glad I did. I totally overlooked that there was a modifier in my code, so all I did was press the **a** key to get those **False** results. As soon as I pressed **Ctrl+a**, it worked fine. This is a great thing to add to issue #712 to help narrow down the cause. It seems that it only works when a modifier is used. That's interesting.

This should fail for you, too, if you use either of the two retCode statements that don't use a modifier:
```python
retCode = keyboard.wait_for_keypress("a")
# retCode = keyboard.wait_for_keypress("a",timeOut=3)
# retCode = keyboard.wait_for_keypress("a",modifiers=["<ctrl>"],timeOut=3)

myResult = dialog.info_dialog(title="Result", message="Exit code: " + str(retCode), width="200")

if retCode:
	myResult
else:
	myResult
```
Even more interestingly, if I use either of the two retCode statements without a modifier, if I press the specified key **twice**, then the exit code is **True**, but I get strange behavior immediately afterwards. When I press the **a** key twice to attempt to satisfy the API call, the API call is satisfied and I get the **True** dialog result. If I then move my cursor to any other location in the text, it starts moving by itself, jumping to the right by two characters each time until I `xkill` AutoKey. I'm going to add this information to that other issue report.

**Comment by sebastiansam55** [comment]
Again, I'm getting the expected behavior, returns false if it hits the timeout, true if the key is pressed.

What version of Autokey are you using? I'm using 0.96.09-beta.9

How are you triggering the script? Can you post the extended logging output?

Note that if you're running the script via the Autokey interface there is a delay of 2 seconds before the script actually starts after hitting the button. 

This is intentional and so you can switch to whatever window you need before it executes, otherwise it'd end up executing against the autokey window which would have focus.

**Comment by Elliria** [comment]
I'm using the **GTK** front-end of **AutoKey 0.95.10** in **Kubuntu 22.04 LTS** (which is kept up-to-date) and I'm triggering the script by pressing the "play" button (that displays the "Run the currently selected script" tooltip when you hover over it) in the AutoKey toolbar. I also add a comment line beneath the script when testing with the press of only one key to make sure my cursor is inside that comment when the script is run to prevent AutoKey from reading the key-press(es).

I'll have to repeat all the tests to give you extended logging output if you'd still like it. Meanwhile, here are some consistently-repeatable tests that give me some consistent and intriguing results:

## No modifier and no timeout:
```python
retCode = keyboard.wait_for_keypress("a")
myResult = dialog.info_dialog(title="Result", message="Exit code: " + str(retCode), width="200")
if retCode:
	myResult
else:
	myResult
```
Result:
* Pressing the specified key **once** displays the **False** exit code. [See the output](https://bpa.st/TVXUQ).
* Pressing the specified key **twice** displays the **True** exit code. [See the output](https://bpa.st/BIKIK).

## Timeout and no modifier:
```python
retCode = keyboard.wait_for_keypress("a",timeOut=3)
myResult = dialog.info_dialog(title="Result", message="Exit code: " + str(retCode), width="200")
if retCode:
	myResult
else:
	myResult
```
Result:
* Pressing the specified key **once** displays the **False** exit code. [See the output](https://bpa.st/LCPBI). <-- *Note that AutoKey did not shut down properly after this and had to be pkilled.*
* Pressing the specified key **twice** displays the **True** exit code. [See the output](https://bpa.st/2LOAI).

## Modifier and no timeout:
```python
retCode = keyboard.wait_for_keypress("a",modifiers=["<ctrl>"])
myResult = dialog.info_dialog(title="Result", message="Exit code: " + str(retCode), width="200")
if retCode:
	myResult
else:
	myResult
```
Result:
* One **short** press of the specified key-combination displays the **False** exit code. [See the output](https://bpa.st/KDTRS).
* One **long** press of the specified key-combination displays the **True** exit code. [See the output](https://bpa.st/XP3ZI).

*Note that in both of these tests, all of the text in the window was selected, indicating that the key-combination was successfully deployed each time, yet only one of them results in a **True** exit code. For me, obtaining a **True** exit code requires non-standard key-pressing behavior that would have to be deliberate.*

## Modifier and timeout:
```python
retCode = keyboard.wait_for_keypress("a",modifiers=["<ctrl>"],timeOut=3)
myResult = dialog.info_dialog(title="Result", message="Exit code: " + str(retCode), width="200")
if retCode:
	myResult
else:
	myResult
```
Result:
* One **short** press of the specified key-combination displays the **False** exit code. [See the output](https://bpa.st/K5K2C).
* One **long** press of the specified key-combination displays the **True** exit code. [See the output](https://bpa.st/QR2HM). <-- *Note that AutoKey did not shut down properly after this and had to be pkilled.*

*Note that in both of these tests, all of the text in the window was selected, indicating that the key-combination was successfully deployed each time, yet only one of them results in a **True** exit code. For me, obtaining a **True** exit code requires non-standard key-pressing behavior that would have to be deliberate.*

**Comment by Elliria** [comment]
> Note that if you're running the script via the Autokey interface there is a delay of 2 seconds before the script actually starts after hitting the button.

I just noticed this. And yes, if I wait a full 2 seconds before pressing the specified key or key-combination, it works.

I guess I'll create an enhancement request for a toggle we can tick to turn that delay off if we know we're going to be running tests inside the main AutoKey window.

**Comment by sebastiansam55** [comment]
It's intended functionality like I said, otherwise it would execute right when you hit the button which is not what most people would use it for, like many things that could be better documented or shown to the user. 

It's a literal `time.sleep(2)` at the link below (for gtkui tho I think it's more or less the same for qtui)
https://github.com/autokey/autokey/blob/master/lib/autokey/gtkui/configwindow.py#L1303

Could definitely add some level of configuration to it tho, would be a good first issue to tackle code wise if you're up to it

You'd basically just replace the `2` in the line linked above with `cm.ConfigManager.SETTINGS[whatever settingname]` and then make sure it's parsed into settings when the file is loaded.

**Comment by sebastiansam55** [comment]
Here it is for the `qtui`;
https://github.com/autokey/autokey/blob/master/lib/autokey/qtui/configwindow.py#L296

The "2000" seconds is the timeout, so you'd have to be sure to multiply in qtui or divide in gtkui so the behavior is consistent

**Comment by Elliria** [comment]
Oh, very cool. Now, that wouldn't be being done on this project, though, right? It's a suggestion for me to do on my own personal build? Also, I opened enhancement issue #828 for this, so hopefully this note will be linked to there so that I can find these steps again.

**Comment by josephj11** [comment]
For reference: There doesn't appear to be a "play" button in the Qt front end. That's a GTK thing. It does have the run script option though.

Noticed that this only works for scripts. Should it work for phrases as well? Especially since phrases can have macros in them that run scripts, etc.

**Comment by Elliria** [comment]
Interesting. Both front-ends have one in AutoKey 0.95.10. The difference is that it's labeled in the Qt front-end and unlabeled in the GTK front-end. Did they change it in AutoKey 0.96.0?

Here's the GTK one:
![AutoKey-GTK-play-button](https://user-images.githubusercontent.com/25792925/229003971-b70f352a-3322-46e4-81e7-7456dba1f72f.png)

And here's the Qt one:
![AutoKey-Qt-play-button](https://user-images.githubusercontent.com/25792925/229003982-89601b4b-f0c1-4eca-bb81-4e2a5f0cdbe4.png)

**Comment by Elliria** [comment]
> Noticed that this only works for scripts. Should it work for phrases as well? Especially since phrases can have macros in them that run scripts, etc.

Good call. It's greyed-out for me when I have phrases open, but you're right. You can run scripts from a phrase, so maybe it should detect that a script call is present in the phrase and be enabled.

**Comment by hfr-hfr** [comment]
The return codes for **`mouse.wait_for_click`** still don't work on my end (autokey-gtk; 0.95.10; X11). While @Elliria 's examples with `keyboard.wait_for_keypress` work as intended, the `mouse.wait_for_click` always returns "None" - whether the event happened or timeout is hit. I guess, the expected behaviour should be the same as `keyboard.wait_for_keypress`, right?

**Comment by Elliria** [comment]
I suppose they should at least give similar results, because neither of them officially provide a return by default. I'm getting the same results as you:

## `keyboard.wait_for_keypress()`
This gives me **True** when I press the keys and **False** when I let it time out:
```python
retCode = keyboard.wait_for_keypress('a',modifiers=['<ctrl>'],timeOut=5)
myResult = dialog.info_dialog(title="Result", message="Exit code: " + str(retCode), width="200")
if retCode:
	myResult # True
else:
	myResult # False
```
## `mouse.wait_for_click()`
This gives me **Exit code: None** when I click and **Exit code: None** when I let it time out:
```python
retCode = mouse.wait_for_click(1,timeOut=3)
myResult = dialog.info_dialog(title="Result", message="Exit code: " + str(retCode), width="200")
if retCode:
	myResult # Exit code: None
else:
	myResult # Exit code: None
```

**Comment by josephj11** [comment]
This is an enhancement request, not a bug report. When I filed it, I didn't know that `keyboard.wait_for_keypress()` returned the correct status because it was (and still is) undocumented.

@Elliria this might be really easy to fix. The one call just requires documentation (via comments in the code to generate it) and the other one might essentially be copying the  return code logic from the one that already works.

**Comment by Elliria** [comment]
True. I added it to my ToDo list.

**Comment by Elliria** [comment]
The underlying issue is a sync disconnect within AutoKey's engine loop. When you trigger `wait_for_keypress`, it halts execution and starts the `self.event.wait` threading-timer. For it to return **True**, a separate background-listener thread needs to catch the keypress, match it to the key that’s being waited for, and call `self.event.set()` to break the wait early. Because that event-trigger is broken or missing in the keyboard-hook backend, the timer never gets notified that the key was pressed. It's, thus, forced to run until the full time-out duration expires, which automatically always triggers a **False** return. Apparently, if you hold the key down for a long-press, it occasionally forces a **True** return, but for normal use, it’s a dead event.

**Comment by josephj11** [comment]
@Elliria Looks like your analysis skills are improving! Is that something you can implement?

**Comment by Elliria** [comment]
Yeeees? Maaaaaaaybe? I kind of stumbled across it while accidentally stumbling across the answer to getting the calendar dialog to work. I suspected that this might be related and might have a similar, simple fix, but it decided to be difficult. Threading is not, by any stretch of the imagination, my wheelhouse since I'm very much a linear programmer (which is a significant weakness). But I'm starting to understand some of it because I've been romping around with it as part of my personal Python play/studies.

Someone very wise recently told me to stop trying to look at the order of execution and, instead, follow the data by asking yourself, "What item does this function need and who or what does it get it from?" with the goal being to trace that item as it gets passed around without worrying about **anything** else that might **also** be going on or **need** to be going on rather than trying to figure out when each of those transfers occur or what the broader implications of each of them are (which is what always trips me up, because the octopus just becomes too confusing for me). It's a way to trick your brain into seeing something non-linear in a linear way rather than trying to understand, let alone memorize, the complex dance of logic that's going on.

That said, I'm still very much stumbling around in the darkness. I sometimes have a flashlight, but still often drop it. That post is a reminder for future me to use as a reference when trying to untangle it (which will likely be easy if the event-trigger exists, but is just broken) ...or... a nudge to any kindly developer who happens to come across it and decides to take a crack at it -- whichever comes first.

**Comment by josephj11** [comment]
@Elliria I haven't done any multithreaded programming, but I did a little ancient pre-web stuff (CICS) which is essentially reentrant code. There were no normal persistent variables, so you always had to embed or pass them to the current instance. The trick is "what is my state _now_?", not how it got that way.

**Comment by Elliria** [comment]
Heh. It's pretty similar. For Python, it's, "what do I want and who gives it to me?"
